src/main/java/org/folio/rest/client/ConfigurationClient.java

```diff
@@ -25,14 +26,55 @@ public class ConfigurationClient extends OkapiClient {
   private static final Logger log = LogManager.getLogger(MethodHandles.lookup().lookupClass());
   private static final String CONFIGURATIONS_URL = "/configurations/entries?query=%s";
   private static final String TLR_SETTINGS_QUERY = "module==\"SETTINGS\" and configName==\"TLR\"";
+  private static final String url = format(CONFIGURATIONS_URL, StringUtil.urlEncode(TLR_SETTINGS_QUERY));
 
   public ConfigurationClient(Vertx vertx, Map<String, String> okapiHeaders) {
     super(vertx, okapiHeaders);
   }
 
+  public Future<TlrSettingsConfiguration> getTlrSettingsOrDefault() {
+    return getSettingsFromStorage(url)
+    .compose(this::tryParseResponseToJson)
+    .compose(res -> {
+      log.debug("getTlrSettingsOrDefault:: Checking for existence of TLR settings");
+      if (res.getConfigs().isEmpty()) {
+        log.debug("getTlrSettingsOrDefault:: TLR settings empty, using default settings");
+        return succeededFuture(new TlrSettingsConfiguration(false, false, null, null, null));
+      } else {
+        log.debug("getTlrSettingsOrDefault:: TLR settings found");
+        return parseSettings(res);
+      }
+    });
+  }
+
   public Future<TlrSettingsConfiguration> getTlrSettings() {
-    String url = format(CONFIGURATIONS_URL, StringUtil.urlEncode(TLR_SETTINGS_QUERY));
+    return getSettingsFromStorage(url)
+      .compose(this::tryParseResponseToJson)
+      .compose(this::parseSettings);
+  }
+
+  private Future<KvConfigurations> tryParseResponseToJson(String settings) {
+    log.info("setting string: " + settings);
+    try{ 
+      return succeededFuture(objectMapper.readValue(settings, KvConfigurations.class));
+    } catch (JsonProcessingException e) {
+      log.error("Failed to parse response: {}", settings);
+      return failedFuture(e);
+    }
+  }
+
+  private Future<TlrSettingsConfiguration> parseSettings(KvConfigurations settings) {
+      return settings
+        .getConfigs().stream()
+        .findFirst()
+        .map(Config::getValue)
+        .map(JsonObject::new)
+        .map(TlrSettingsConfiguration::from)
+        .map(Future::succeededFuture)
+        .orElseGet(() -> failedFuture("Failed to find TLR configuration"));
```

Comment:
Actually, I think the whole difference between the two cases boils down to the return value of this line. If TLR-configuration was fethed successfully, but the response is empty, this line should return:
- a failed future if we don't want the default value
- a succeeded future with the default value otherwise
In this case we can extract this supplier into a method parameter, and leave the rest of the code intact. Do you think this could work?

```
public Future<TlrSettingsConfiguration> getTlrSettingsOrDefault() {
    return getTlrSettings(
      () -> succeededFuture(new TlrSettingsConfiguration(false, false, null, null, null)));
  }

  public Future<TlrSettingsConfiguration> getTlrSettings() {
    return getTlrSettings(() -> failedFuture("Failed to find TLR configuration"));
  }

  private Future<TlrSettingsConfiguration> getTlrSettings(
    Supplier<Future<TlrSettingsConfiguration>> otherwise) {

    String url = format(CONFIGURATIONS_URL, StringUtil.urlEncode(TLR_SETTINGS_QUERY));

    return okapiGet(url)
      .compose(response -> {
        int responseStatus = response.statusCode();
        if (responseStatus != 200) {
          String errorMessage = format("Failed to find TLR configuration. Response: %d %s",
            responseStatus, response.bodyAsString());
          log.error(errorMessage);
          return failedFuture(new HttpException(GET, url, response));
        } else {
          try {
            return objectMapper.readValue(response.bodyAsString(), KvConfigurations.class)
              .getConfigs().stream()
              .findFirst()
              .map(Config::getValue)
              .map(JsonObject::new)
              .map(TlrSettingsConfiguration::from)
              .map(Future::succeededFuture)
              .orElseGet(otherwise);
          } catch (JsonProcessingException e) {
            log.error("Failed to parse response: {}", response.bodyAsString());
            return failedFuture(e);
          }
        }
      });
  }
```

Reply:
I'll plug it in and test it.